File: Binary Search/278. First Bad Version.cpp

// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int l = 1, r = n;
        while(l<r){
            int mid = l + (r-l)/2;
            if(isBadVersion(mid)){
                r = mid;
            }else{
                l = mid+1;
            }
        } 
        return l;
    }
};




bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int l = 1, r = n;
        while(l<=r){
            int mid = l + (r-l)/2;
            if(isBadVersion(mid)){
                r = mid-1;
            }else{
                l = mid+1;
            }
        } 
        return l;
    }
};




class Solution {
public:
    int firstBadVersion(int n) {
        int start=0, end=n;
        while(end-start>1){
            int mid=start+(end-start)/2;
            if(isBadVersion(mid))  end=mid;
            else  start=mid;
        }
        return end;
    }
};

File: Binary Search/278. First Bad Version.py

class Solution:
    def firstBadVersion(self, n):
        return self.binarySearch(1, n)
    
    def binarySearch(self, low, high):
        if low == high:
            return low
        mid = (low + high)//2
        if isBadVersion(mid) == True:
            return self.binarySearch(low, mid)
        else:
            return self.binarySearch(mid + 1, high)



class Solution:
       def firstBadVersion(self, n):
        return bisect.bisect(type('', (), {'__getitem__': lambda self, i: isBadVersion(i)})(), False, 0, n)


import bisect
class Solution():
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        self.__getitem__ = isBadVersion
        return bisect.bisect_left(self, True, 1, n)

class Solution:
    def firstBadVersion(self, n):
        class Wrap:
            def __getitem__(self, i):
                return isBadVersion(i)
        return bisect.bisect(Wrap(), False, 0, n)